src/cudadev/CUDACore/ScopedContext.cc

```diff
@@ -81,8 +54,22 @@ namespace cms::cuda {
     }
 
     void ScopedContextHolderHelper::enqueueCallback(int device, cudaStream_t stream) {
-      cudaCheck(
-          cudaStreamAddCallback(stream, cudaScopedContextCallback, new CallbackData{waitingTaskHolder_, device}, 0));
+      SharedEventPtr event = getEventCache().get();
+      cudaEventRecord(event.get(), stream);
+      edm::async(waitingTaskHolder_, [device, stream, event = std::move(event)]() mutable {
+        auto status = cudaEventSynchronize(event.get());
+        event.reset();
+        if (status != cudaSuccess) {
+          auto error = cudaGetErrorName(status);
+          auto message = cudaGetErrorString(status);
+          throw std::runtime_error(
+              "Callback of CUDA stream " + std::to_string(reinterpret_cast<unsigned long>(stream)) + " in device " +
+              std::to_string(device) + " error " + std::string(error) + ": " + std::string(message));
+        } else {
+          //std::cout << " GPU kernel finished (in callback) device " << device << " CUDA stream "
+          //          << stream << std::endl;
+        }
```

Comment:
This could simply be
```suggestion
        cudaCheck(cudaEventSynchronize(event.get()));
```
except for the mention of the callback in the error message ?

Reply:
I agree, I wasn't sure if the "Callback on device X" would be useful information or not, so I decided to keep the message from the `cudaStreamAddCallback()` version. Ok, reminding myself of `cudaCheck()` again I could use the `description` argument for that information. In that case I'd be tempted to change the `description` to `std::string_view` (following similar change in CMSSW). On the other hand, using `description` argument would imply (unnecessarily) constructing the error message also for the successful cases